src/Data/Singletons/Single/Data.hs

```diff
@@ -172,12 +170,6 @@ singCtor dataName (DCon _tvbs cxt name fields rty)
                       DRecC [ (singValName field_name, noBang, arg)
                             | (field_name, _, _) <- rec_fields
                             | arg <- args ]
-  return $ DCon tvbs
-                []
-                sName
-                conFields
-                (DConT (singTyConName dataName) `DAppT` foldType pCon indices)
-  where buildArgType :: OSet Name -> DType -> DType -> SgM DType
-        buildArgType bound_kvs ty index = do
-          (ty', _, _, _, _, _) <- singType bound_kvs index ty
```

Comment:
I realized that this code was incredibly overwrought, considering the extremely simple purpose it served: take an argument type, promote it, and slap a `Sing` in front of it. That's all. We were using `singType` to accomplish this without much good reason, since it performs a whole cavalcade of extra stuff that is necessary for top-level type signatures, but unnecessary for argument types (as argument types don't have `forall`s, contexts, etc.) This is even more glaring when you consider that we have to discard five (!) results from `singType`.

We now accomplish the same thing in a much more direct fashion—see `args` above. Technically, this code _hasn't_ become pure, because we still need to promote the argument types before slapping `Sing` in front of them. It's just that the impurity is in a different location than before.